File: dsl/src/test/java/dev/kolibrium/demo/dsl/java/ProductsTest.java
package dev.kolibrium.demo.dsl.java;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.Cookie;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeDriverService;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.io.File;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Duration;
import java.util.List;
import java.util.function.Predicate;

import static org.assertj.core.api.Assertions.assertThat;
import static org.openqa.selenium.chromium.ChromiumDriverLogLevel.INFO;

// example from https://www.selenium.dev/documentation/test_practices/encouraged/page_object_models/#page-component-objects

public class ProductsTest {
    private WebDriver driver;

    @BeforeEach
    public void setUp() {
        var driverService = new ChromeDriverService.Builder()
                .withLogFile(new File("chrome.log"))
                .withLogLevel(INFO)
                .withReadableTimestamp(true)
                .build();

        var options = new ChromeOptions();
        options.addArguments("--incognito");
        options.addArguments("--start-maximized");
        options.setExperimentalOption("excludeSwitches",
                List.of("enable-automation"));

        driver = new ChromeDriver(driverService, options);
        driver.get("https://www.saucedemo.com");
    }

    @AfterEach
    public void tearDown() {
        driver.quit();
    }

    @Test
    @DisplayName("there should be 6 products in total")
    public void testProductInventory() {
        var productsPage = new ProductsPage(driver);
        var products = productsPage.getProducts();
        assertThat(products.size()).isEqualTo(6);
    }
    
    @Test
    @DisplayName("prices should be correctly displayed")
    public void testProductPrices() {
        var productsPage = new ProductsPage(driver);

        // Pass a lambda expression (predicate) to filter the list of products
        // The predicate or "strategy" is the behavior passed as parameter
        var backpack = productsPage.getProduct(p -> p.getName().contains("Backpack")); // page component object
        var bikeLight = productsPage.getProduct(p -> p.getName().contains("Bike Light"));

        assertThat(backpack.getPrice()).isEqualTo(new BigDecimal("29.99"));
        assertThat(bikeLight.getPrice()).isEqualTo(new BigDecimal("9.99"));
    }
}

class ProductsPage {
    private final WebDriver driver;

    private final By headerContainer = By.className("header_container");
    private final By name = By.className("inventory_item");

    public ProductsPage(WebDriver driver) {
        this.driver = driver;
        var cookie = new Cookie("session-username", "standard_user");
        driver.manage().addCookie(cookie);
        driver.get("https://www.saucedemo.com/inventory.html");

        // No assertions, throws an exception if the element is not loaded
        new WebDriverWait(driver, Duration.ofSeconds(3))
                .until(d -> d.findElement(headerContainer));
    }

    // Returning a list of products is a service of the page
    public List<Product> getProducts() {
        return driver.findElements(name)
                .stream()
                .map(Product::new) // Map WebElement to a product component
                .toList();
    }

    // Return a specific product using a boolean-valued function (predicate)
    // This is the behavioral Strategy Pattern from GoF
    public Product getProduct(Predicate<Product> condition) {
        return getProducts()
                .stream()
                .filter(condition) // Filter by product name or price
                .findFirst()
                .orElseThrow();
    }
}

class Product {
    private final WebElement root;

    private final By name = By.className("inventory_item_name");
    private final By price = By.className("inventory_item_price");
    private final By id = By.id("add-to-cart-sauce-labs-");

    // The root element contains the entire component
    public Product(WebElement root) {
        this.root = root;
    }

    public String getName() {
        // Locating an element begins at the root of the component
        return root.findElement(name).getText();
    }

    public BigDecimal getPrice() {
        return new BigDecimal(
                root.findElement(price)
                        .getText()
                        .replace("$", "")
        ).setScale(2, RoundingMode.UNNECESSARY); // Sanitation and formatting
    }

    public void addToCart() {
        root.findElement(id).click();
    }
}

File: dsl/src/test/kotlin/dev/kolibrium/demo/dsl/_03/ProductsTest.kt
package dev.kolibrium.demo.dsl._03

import dev.kolibrium.dsl.selenium.creation.Arguments.Chrome.incognito
import dev.kolibrium.dsl.selenium.creation.Arguments.Chrome.start_maximized
import dev.kolibrium.dsl.selenium.creation.Switches.enable_automation
import dev.kolibrium.dsl.selenium.creation.chromeDriver
import dev.kolibrium.dsl.selenium.interactions.cookies
import dev.kolibrium.selenium.className
import dev.kolibrium.selenium.classNames
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement
import org.openqa.selenium.chromium.ChromiumDriverLogLevel.INFO
import java.math.BigDecimal
import java.math.RoundingMode

class ProductsTest {
    private lateinit var driver: WebDriver

    @BeforeEach
    fun setUp() {
        driver = chromeDriver {
            driverService {
                logFile = "chrome.log"
                logLevel = INFO
                readableTimestamp = true
            }
            options {
                arguments {
                    +incognito
                    +start_maximized
                }
                experimentalOptions {
                    excludeSwitches {
                        +enable_automation
                    }
                }
            }
        }.apply {
            get("https://www.saucedemo.com")
        }
    }

    @AfterEach
    fun tearDown() {
        driver.quit()
    }

    @Test
    fun `there should be 6 products in total`() {
        with(driver) {
            with(ProductsPage()) {
                val products = getProducts()
                products.size shouldBe 6
            }
        }
    }

    @Test
    fun `prices should be correctly displayed`() {
        with(driver) {
            with(ProductsPage()) {
                val backpack = getProduct { it.name.text.contains("Backpack") }
                val bikeLight = getProduct { it.name.text.contains("Bike Light") }

                backpack.getPrice() shouldBe BigDecimal("29.99")
                bikeLight.getPrice() shouldBe BigDecimal("9.99")
            }
        }
    }
}

context(WebDriver)
class ProductsPage {
    private val headerContainer by className("header_container")
    private val names by classNames("inventory_item")

    init {
        cookies {
            addCookie(name = "session-username", value = "standard_user")
        }
        get("${currentUrl}inventory.html")
        check(headerContainer.isDisplayed) {
            "This is not the Inventory Page, current page is: " + this@WebDriver.currentUrl
        }
    }

    fun getProducts() = names.map { webElement ->
        Product(webElement)
    }

    fun getProduct(condition: (Product) -> Boolean): Product {
        return getProducts()
            .asSequence()
            .filter(condition) // Filter by product name or price
            .firstOrNull()
            ?: throw NoSuchElementException()
    }
}

class Product(val root: WebElement) {
    val image by root.className("inventory_item_img")
    val name by root.className("inventory_item_name")
    val description by root.className("inventory_item_desc")
    private val price by root.className("inventory_item_price")

    fun getPrice(): BigDecimal {
        return BigDecimal(price.text.replace("$", "")).setScale(
            2,
            RoundingMode.UNNECESSARY
        ) // Sanitation and formatting
    }
}
